system/service/auth_credentials.go

```diff
@@ -0,0 +1,1157 @@
+	var (
+		hash []byte
+	)
+
+	if hash, err = svc.hashPassword(password); err != nil {
+		return
+	}
+
+	if err = svc.removePasswordCredentials(ctx, userID); err != nil {
+		return
+	}
+
+	// Add new credentials with new password
+	c := &types.Credentials{
+		ID:          nextID(),
+		CreatedAt:   *now(),
+		OwnerID:     userID,
+		Kind:        credentialsTypePassword,
+		Credentials: string(hash),
+	}
+
+	return store.CreateCredentials(ctx, svc.store, c)
+}
+
+//CheckPassword verifies if password matches any of the valid credentials
+//func (svc *auth) CheckPassword(cc types.CredentialsSet, password string) bool {
+//	return findMatchingCredentials(cc, password, true) != nil
+//}
+
+// SetPassword sets new password for a user
+//
+// This function also records an action
+//
+// this method is used in 2 scenarios:
+//
+//  SELF:
+//    user forgot the password and needs to reset it
+//    there should be protocols prior to this point that
+//    authenticate and validate users
+//
+//  USER MANAGEMENT:
+//    administrator is resetting password for another user
+//
+func (svc *auth) SetPassword(ctx context.Context, userID uint64, password string) (err error) {
+	var (
+		u  *types.User
+		cc types.CredentialsSet
+
+		aam = &authActionProps{
+			user:        u,
+			credentials: &types.Credentials{Kind: credentialsTypePassword},
+		}
+	)
+
+	err = func() error {
+		if !svc.settings.Auth.Internal.Enabled {
+			return AuthErrInternalLoginDisabledByConfig(aam)
+		}
+
+		if !svc.CheckPasswordStrength(password) {
+			return AuthErrPasswordNotSecure(aam)
+		}
+
+		u, err = store.LookupUserByID(ctx, svc.store, userID)
+		if errors.IsNotFound(err) {
+			return AuthErrPasswordChangeFailedForUnknownUser(aam)
+		}
+
+		aam.setUser(u)
+		ctx = internalAuth.SetIdentityToContext(ctx, u)
+
+		cc, _, err = store.SearchCredentials(ctx, svc.store, types.CredentialsFilter{
+			Kind:    credentialsTypePassword,
+			OwnerID: userID,
+			Deleted: filter.StateInclusive})
+
+		if err != nil {
+			return err
+		}
+
+		if isPasswordReused(cc, password, svc.opt.PasswordReuseTimeWindow) {
+			return AuthErrPasswordSetFailedReusedPasswordCheckFailed(aam)
+		}
+
+		if err != svc.SetPasswordCredentials(ctx, userID, password) {
+			return err
+		}
+
+		return nil
+	}()
+
+	return svc.recordAction(ctx, aam, AuthActionChangePassword, err)
+}
+
+// PasswordSet checks and returns true if user's password is set
+// False is also returned in case user does not exist.
+func (svc *auth) PasswordSet(ctx context.Context, email string) (is bool) {
+
+	//svc.settings.Auth.External.Enabled
+	u, err := store.LookupUserByEmail(ctx, svc.store, email)
+	if err != nil {
+		return
+	}
+
+	cc, _, err := store.SearchCredentials(ctx, svc.store, types.CredentialsFilter{
+		OwnerID: u.ID,
+		Kind:    credentialsTypePassword,
+	})
+	if err != nil {
+		return
+	}
+
+	if len(cc) > 0 && svc.settings.Auth.Internal.Enabled {
```

Comment:
you can move the `svc.settings.Auth.Internal.Enabled` bit to the top and avoid needless computation in case it is not enabled